### PR TITLE
[embedded] Introduce class-bound existentials into Embedded Swift

### PR DESCRIPTION
**Pull request by kubamracek**
Motivated by need for protocol-based dynamic dispatch, which hasn't been possible in Embedded Swift due to a full ban on existentials. This lifts that restriction but only for class-bound existentials: Class-bound existentials are already (even in desktop Swift) much more lightweight than full existentials, as they don't need type metadata, their containers are typically 2 words only (reference + wtable pointer), don't incur copies (only retains+releases).

Included in this PR:
- [x] Non-generic class-bound existentials, executable tests for those.
- [x] Extension methods on protocols and using those from a class-bound existential.
- [x] RuntimeEffects now differentiate between Existential and ExistentialClassBound.
- [x] PerformanceDiagnostics don't flag ExistentialClassBound in Embedded Swift.
- [x] WTables are generated in IRGen when needed.

Left for follow-up PRs:
- [ ] Generic classes support

### CODE REVIEW

File: lib/SILOptimizer/PassManager/PassPipeline.cpp [hunk 1]
```diff
@@ -457,7 +457,13 @@ void addFunctionPasses(SILPassPipelinePlan &P,
   P.addMem2Reg();
 
   // Run the existential specializer Pass.
-  P.addExistentialSpecializer();
+  if (!P.getOptions().EmbeddedSwift) {
+    // MandatoryPerformanceOptimizations already took care of all specializations
+    // in embedded Swift mode, running the existential specializer might introduce
+    // more generic calls from non-generic functions, which breaks the assumptions
+    // of embedded Swift.
+    P.addExistentialSpecializer();
```
Comment (kubamracek) [inline lib/SILOptimizer/PassManager/PassPipeline.cpp:465]: @eeckstein Any better ideas what to do here? Or is this reasonable to keep? Temporarily?

File: lib/SIL/Utils/InstructionUtils.cpp [hunk 2]
```diff
@@ -963,7 +961,7 @@ RuntimeEffect swift::getRuntimeEffect(SILInstruction *inst, SILType &impactType)
       rt |= RuntimeEffect::ObjectiveC | RuntimeEffect::MetaData;
       break;
     case SILFunctionTypeRepresentation::WitnessMethod:
-      rt |= RuntimeEffect::MetaData | RuntimeEffect::Existential;
+      rt |= RuntimeEffect::MetaData; // ???
```
[the remark this answers is not in the document]
Comment (eeckstein) [reply lib/SIL/Utils/InstructionUtils.cpp:964]: yes, I think this makes sense

**Review by eeckstein — APPROVED** [review]
nice!

File: lib/SIL/Utils/InstructionUtils.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kubamracek) [reply lib/SIL/Utils/InstructionUtils.cpp:964]: Right, but do you have any hints on how to implement this? We have an ApplySite and an ApplyInst here, how do we go from that to the protocol that's in used?

[the remark this answers is not in the document]
Comment (eeckstein) [reply lib/SIL/Utils/InstructionUtils.cpp:964]: You could do it as in `hasValidSignatureForEmbedded`: check if all generic parameters are class bound. The function type you can get from `ApplySite::getOrigCalleeType()`

[the remark this answers is not in the document]
Comment (eeckstein) [reply lib/SIL/Utils/InstructionUtils.cpp:964]: It's even simpler: you can use `SILFunctionType::getWitnessMethodConformanceOrInvalid`

[the remark this answers is not in the document]
Comment (kubamracek) [reply lib/SIL/Utils/InstructionUtils.cpp]: Ok, done!

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test Linux platform

**Pull request merged**